bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/scheduler/AbstractExpression.java

```diff
@@ -207,22 +207,51 @@ protected void prune() {
 
     @Override
     public Date getTimeAfter(Date afterTime) {
+
+        Date currentStartDate = getStartDate();
+
+        if (hasFloatingStartDate()) {
+            try {
+                clearCandidates();
+                setStartDate(afterTime);
+            } catch (IllegalArgumentException | ParseException e) {
+                logger.error("An exception occurred while setting the start date : '{}'", e.getMessage());
+            }
+        }
+
         if (getCandidates().isEmpty()) {
             try {
                 setStartDate(afterTime);
-                parseExpression(expression);
             } catch (ParseException e) {
-                logger.error("An exception occurred while parsing the expression : '{}'", e.getMessage());
+                logger.error("An exception occurred while setting the start date : '{}'", e.getMessage());
             }
         }
 
         if (!getCandidates().isEmpty()) {
-
-            Collections.sort(getCandidates());
-
-            for (Date candidate : getCandidates()) {
-                if (candidate.after(afterTime)) {
-                    return candidate;
+            if (getCandidates().size() == 1) {
+                return getCandidates().get(0);
+            } else {
+                while (getCandidates().size() > 1) {
```

Comment:
I comment the else branch, so let's have a look at the code in that branch only (`getCandidates().size() == 1` is checked outside of the else branch).

```java
while (getCandidates().size() > 1) {

    Collections.sort(getCandidates());

    Date newStartDate = null;

    try {

        for (Date candidate : getCandidates()) {
            newStartDate = candidate;
            if (candidate.after(afterTime)) {
                setStartDate(currentStartDate);
                return candidate;
            }
        }

        clearCandidates();
        setStartDate(newStartDate);
    } catch (IllegalArgumentException | ParseException e) {
        logger.error("An exception occurred while parsing the expression : '{}'", e.getMessage());
    }
}
```

If the last call to `setStartDate(newStartDate);` results in a candidate list of size 1, the condition if false and the while loop body is not executed again. Shouldn't also the one remaining candidate be checked if it is after the after time and returned if this is true.